### ISSUE TITLE
Question about trainable parameters

### ISSUE THREAD

**Issue by Mu-Y** [kyutai-labs/moshivis#6]
Hello,

I have a question about what parameters are trainable in MoshiVis.  In the paper, Figure 3 caption stated that "In practice, we only train the first two audio streams for speech samples. This allows for faster training as we need fewer parallel calls to the depth transformer. " Does this mean you also make the embedding layers for the first 2 audio stream tokens trainable, in addition to the shared Cross-Attention parameters?  Why does it "needs fewer parallel calls to the depth transformer"?  Any other trainable parameters in MoshiVis except the inserted Cross-Attention modules? 

Thanks!

**Comment by ameroyer** [comment]
Hello,

**fewer parallel calls to the depth transformer**
At inference, the depth transformer outputs $K$ audio streams in an autoregressive manner, starting from the output of the speech-LLM backbone. At training, these $K$ depth transformer calls are done in parallel using the ground-truth audio streams of the training data as inputs (similar to standard auto-regressive training), hence the fewer "parallel calls" since we only compute the outputs of a subset of the audio streams to use in the loss for MoshiVis.


**audio embedding layers + trainable parameters**
Yes indeed, the only trainable parameters in the paper and demo models are the gated cross-attention modules (QKV projections, output projection and layer norm + gating module), and at the output of the image encoder we also have a small trainable adapter module (single linear layer + layer norm). 

We only use the depth transformer output of the first two audio streams as signal in  the loss but do not train any of the audio embeddings or other audio-related parameters.

Hope that clarifies everything!

**Comment by Mu-Y** [comment]
Thanks for the fast response!

So during training, for "speech samples", is loss computed from the text stream + the first 2 audio streams, or only the first 2 audio streams? My understanding is that this only applies to assistant streams. In this case, for user audio streams, do you use all streams? And during inference, all assistant streams are predicted? I'm wondering why this approach could accurately predict the remaining audio streams, since the remaining audio streams are not used in training.  

Also, for "speechless samples", is loss only computed over the assistant text span, or computed over the text spanning both assistant and user?  Thanks!

**Comment by ameroyer** [comment]
Hello,

Sorry for the late reply,

- Yes, during training the loss for speech samples, is applied only on the text stream + first 2 audio streams of the assistant. We actually usually don't even apply loss on the user audio streams; Similarly for speechless samples we only apply the loss on the assistant's text span.
- At inference all audio streams of the assistant are decoded
- As for why only partially training the audio streams works so well; This is a good question, my main insights on this are:
  - The speech LLM is already trained and frozen. In particular, since the depth transformer has an auto-regressive nature, if we're able to keep the first audio codebooks perfectly in distribution, this should make the weights of the later depformer levels still able to decode them properly.
  - We're not trying to add any new audio capabilities per-say, only new text/semantics behaviour; In a way the loss on the audio streams is mainly a way to regularise the model and prevent the output audio to go out of distribution when it sees image inputs.
  - Nevertheless, we also experimented with fine-tuning all the audio streams, which can slightly improve audio quality but with higher cost of training; And conversely, going to the extreme opposite of only training the text tokens yields unintelligible audio
